### PR TITLE
Timeout for base llms

### PR DESCRIPTION
**Pull request by alessarcross**
Description: Adding timeout as an option to the BaseLLM class.

### CODE REVIEW

File: langchain/llms/base.py [hunk 1]
```diff
@@ -255,19 +257,45 @@ async def _agenerate_helper(
...
-                else await self._agenerate(prompts, stop=stop)
+                else self._agenerate(prompts, stop=stop)
+            )
+            task = asyncio.create_task(coroutine)
+            tasks_to_await = {
+                task,
+                asyncio.sleep(timeout) if timeout else asyncio.Future(),
+                abort_signal if abort_signal else asyncio.Future(),
+            }
+            done, pending = await asyncio.wait(
```
Comment (dev2049) [inline langchain/llms/base.py:281]: not very familiar with asyncio but looks like you can pass timeout directly to wait, maybe cleaner to use that?

File: langchain/llms/base.py [hunk 2]
```diff
@@ -255,19 +257,45 @@ async def _agenerate_helper(
...
                     stop=stop,
                     run_manager=run_managers[0] if run_managers else None,
                     **kwargs,
                 )
                 if new_arg_supported
-                else await self._agenerate(prompts, stop=stop)
+                else self._agenerate(prompts, stop=stop)
+            )
+            task = asyncio.create_task(coroutine)
+            tasks_to_await = {
```
Comment (dev2049) [inline langchain/llms/base.py:276]: ooc (out of curiosity) any reason to make this a set vs just a list/tuple?

File: langchain/llms/base.py [hunk 3]
```diff
@@ -82,6 +82,8 @@ class BaseLLM(BaseLanguageModel, ABC):
     callback_manager: Optional[BaseCallbackManager] = Field(default=None, exclude=True)
     tags: Optional[List[str]] = Field(default=None, exclude=True)
     """Tags to add to the run trace."""
+    timeout: Optional[float] = Field(default=None, exclude=True)
```
[the remark this answers is not in the document]
Comment (dev2049) [reply langchain/llms/base.py:85]: looks like for agents we call this max_execution_time

**Review by dev2049 — COMMENTED** [review]
not v familiar with asyncio but high level llm part makes sense to me. would be great if @agola11 or @nfcampos could take a look

File: langchain/llms/base.py [hunk 4]
```diff
@@ -255,19 +257,45 @@ async def _agenerate_helper(
         stop: Optional[List[str]],
         run_managers: List[AsyncCallbackManagerForLLMRun],
         new_arg_supported: bool,
+        timeout: Optional[float] = None,
```
[the remark this answers is not in the document]
Comment (alessarcross) [reply langchain/llms/base.py:260]: sounds good 👍

File: langchain/llms/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alessarcross) [reply langchain/llms/base.py:85]: sg - i'll use the already existing params then

File: langchain/llms/base.py (hunk 2 above)
Comment (alessarcross) [reply langchain/llms/base.py:274]: list tuple works too, just wanted to eliminate possibility of duplicate tasks ever (which is unlikely to happen tho even if we use a list)

File: langchain/llms/base.py (hunk 1 above)
Comment (alessarcross) [reply langchain/llms/base.py:281]: sure - just cleaned it up 👍

File: langchain/llms/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alessarcross) [reply langchain/llms/base.py]: done ✅

File: langchain/llms/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alessarcross) [reply langchain/llms/base.py]: done

**Comment by leo-gan** [comment]
@alessarcross Hi , could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

**Comment by efriis** [comment]
Closing because the PR wouldn't line up with the current directory structure of the library (would need to be in /libs/langchain/langchain instead of /langchain). Feel free to reopen against the current head if it's still relevant!

**Pull request closed without merging**